### ISSUE TITLE
Magento loading the wrong Tax info on admin Invoice and Creditmemo page

### ISSUE THREAD

**Issue by srenon** [magento/magento2#31197]
After upgrading from Magento 2.3.3 to 2.4.1, we are having random broken invoice pages. After further debugging, this seems to be core issue.

### Preconditions (*)

1. Magento 2.4-develop

### Steps to reproduce (*)

1. Install clear Magento!
2. Set-up tax rule for `CA` region.
3. Go to `Store`->`Configuration`->`Sales`->`Tax`->`Orders, Invoices, Credit Memos Display Settings` and set
     `Include Tax In Order Total` -> `Yes`;
     `Display Full Tax Summary` -> `Yes`;
4. Create a simple product.
5. Place order with this product chose `State/Province` -> `California`.
6. Place one more order and choose any `State/Province` except for `California` (to avoid taxes).
7. Open the second order on the admin panel and create an invoice for it.
8. Open the second order again, go to the `Invoices` tab, and open the invoice.

### Expected result (*)

1. Invoice page loaded successfully

### Actual result (*)

1. Invoice page broken
![image](https://user-images.githubusercontent.com/51680850/102873281-98b17480-4449-11eb-9aff-958e654e9ec8.png)

### Additional information

file

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Sales/Block/Adminhtml/Order/Totals/Tax.php#L98

[app/code/Magento/Tax/Model/ResourceModel/Sales/Order/Tax/Collection.php](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Tax/Model/ResourceModel/Sales/Order/Tax/Collection.php#L36)

![image](https://user-images.githubusercontent.com/1415141/101444321-3d31a380-38ed-11eb-8fa3-ebbd7ff1a0ce.png)

In the image above ``order->getId()`` is actually an instance of Magento\Sales\Model\Order\Creditmemo or Magento\Sales\Model\Order\Invoice therefore ``getId()`` will return the invoice or credit memo id instead of the order id which will send the wrong information to the SQL below.

```
SELECT 
  `main_table`.* 
FROM 
  `sales_order_tax` AS `main_table` 
WHERE 
  (main_table.order_id = 1) 
ORDER BY 
  `process` ASC
```

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by chequille** [comment]
Hi Guys, do have the same problem. The Page showing the invoice as described above breaks.
It happens if correct tax is 0.00 f.e. if a coupon code was entered and nothing has to be paid.
It happens as well if a product without tax is the only product in the order.
So as this worked everything in 2.3.5 I hope you will fix this as fast as possible.
Everybody is recommending updating to 2.4.1 but new bugs prevents us all from doing it.
Looking forward to hearing from you.
Best regards,
Chequille

**Comment by srenon** [comment]
@chequille You could use a composer patch to temporarily fix this issue until a new version is released. See https://devdocs.magento.com/guides/v2.4/comp-mgr/patching/composer.html

filename: {root}/composer.json

    "magento/module-sales": {
          "GITHUB:Magento2 - 31197 - Loading wrong order tax info":"patches/composer/magento-github-issue-31197.diff"
    }

filename: {root}/patches/composer/magento-github-issue-31197.diff

    Index: vendor/magento/module-sales/Block/Adminhtml/Order/Totals/Tax.php
    ===================================================================
    diff --git a/vendor/magento/module-sales/Block/Adminhtml/Order/Totals/Tax.php b/vendor/magento/module-sales/Block/Adminhtml/Order/Totals/Tax.php
    --- a/vendor/magento/module-sales/Block/Adminhtml/Order/Totals/Tax.php	(date 1607408606000)
    +++ b/vendor/magento/module-sales/Block/Adminhtml/Order/Totals/Tax.php	(date 1607408606000)
    @@ -95,7 +95,7 @@
 
             $taxClassAmount = $this->_taxHelper->getCalculatedTaxes($source);
             if (empty($taxClassAmount)) {
    -            $rates = $this->_taxOrderFactory->create()->getCollection()->loadByOrder($source)->toArray();
    +            $rates = $this->_taxOrderFactory->create()->getCollection()->loadByOrder($this->getOrder())->toArray();
             $taxClassAmount = $this->_taxCalculation->reproduceProcess($rates['items']);
             }

**Comment by chequille** [comment]
@srenon
Thanks for this. I made the changes manually. It is only one line to change :)
After chaning the page break when showing an invoice with zero tax did not happen again. So far so good.
But unfortunately, I found another bug regarding tax. It seems in 2.4.1 there are a lot of changes which did not solve problems but made new problems. I will show you what I found:
1. Tax calculation is set in my system to "afterdiscount"
2. I have a coupon for a special product at 49,-€ which reduces this amount 100% and no delivery costs are applied. Therefore, customer does not have to pay something. The invoice total and tax must be ZERO
3. This working perfectly in my 2.3.5 store which is online

No, I show ou with some screenshots what happens in 2.4.1
I placed an order with a coupon code and the product I mentioned above.
Bank transfer payment, so invocie is not generated.
In the backend the order shows the following totals and tx amount which are correct, all zero !!!!
![grafik](https://user-images.githubusercontent.com/15807794/101965581-77fd4980-3c15-11eb-83c4-af4c14388572.png)

So far so good, but now I generate the invoice for this order and after doing so, we suddenly have a non zero toal and a none zero tax, and this is wrong!!!!
![grafik](https://user-images.githubusercontent.com/15807794/101965735-fe199000-3c15-11eb-8818-dfb3f8e49cd0.png)

This is a very strange bug and I hope this will be solved very soon, as well with some patches, because waiting for the next release is probably to long.

Looking forward to hearing from you.

Best regards,
Chequille

**Comment by srenon** [comment]
@chequille .... did you open a new ticket for the issue mention above (I have not fully reviewed but it seems to be a different issue)?

**Comment by chequille** [comment]
@srenon
No, I did not, because I thought this is also related to the wrong tax calculation as it is as well in this ticket.
On the other hand, as it happens when generating the invoice it is another issue.

Could you transfer it to a new one?

**Comment by srenon** [comment]
@chequille ... I only have limited access to this repo... so please go ahead a create a new ticket.

**Comment by chequille** [comment]
@srenon
Ok, did it, see #31366

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-40113` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Hi @srenon. Thank you for your report. 
The issue has been fixed in magento/magento2#31198 by @srenon in 2.4-develop branch
Related commit(s):
 - [41c3aa20144741faa1a5935c1764416872166474](https://github.com/magento/magento2/commit/41c3aa20144741faa1a5935c1764416872166474)
 - [4b007c23916f94a1dd4acd3ebcb7d75fa2daf47b](https://github.com/magento/magento2/commit/4b007c23916f94a1dd4acd3ebcb7d75fa2daf47b)
 - [0e58811126330a518ecf609a7ce98dd7c36f1484](https://github.com/magento/magento2/commit/0e58811126330a518ecf609a7ce98dd7c36f1484)
 - [84b1ea0f3ff98c8e57049644fd1f456804738d0f](https://github.com/magento/magento2/commit/84b1ea0f3ff98c8e57049644fd1f456804738d0f)

The fix will be available with the upcoming 2.4.3 release.
